### ISSUE TITLE
Allow shipcat and charts share deployment kind logic

### ISSUE THREAD

**Issue by clux** [babylonhealth/shipcat#382]
Currently magic properties either have certain assumptions (can't use kong with elb charts), and cause us to not have to do certain tests (no health check necessary on headless services), some cronjob only setups still require a deployment which causes a debian with `sleep: Inifinity` running using ~0 resources. Having the concepts of deployment vs. statefulset deploy makes this worse by having shipcat to know how to wait for upgrades / get logs / port-forward / shell into a pod.

These things can easily be fixed if we introduced the concept of **kinds** into shipcat. Here are some __preliminary__ suggested kinds:

### Deployment
Standard. Deployment only. Optional extra cronjobs, optional extra worker deployments.
Can't use volumeclaims. Must have a healthcheck.

### Stateful
Same as above, except all main properties are shunted to a stateful set. Required for persistent volume claims.

### Headless
Same as deployment, but can not have a healthcheck.

### Cron
No deployment at all. Just cronjobs. Can have volume claims.

**Comment by janslow** [comment]
A couple of ideas

# `mainWorkload` property
Introduce a `mainWorkload` key which contains all of the properties which only affect the deployment.

```yaml
# services/foo/manifest.yml
mainWorkload:
	# enabled: true # set to false for headless/cron types
	# kind: deployment # or use kind: stateless
	image: ...
	resources: ...
	env: ...
workers:
	foo-worker: ...
cronJobs:
	foo-hourly: ...
```

# `workloads` property
Rather than having multiple fields for all types of workloads, use a single object:

```yaml
# services/foo/manifest.yml
mainWorkload:
	# enabled: true # set to false for headless/cron types
	# kind: deployment # or use kind: stateless
	image: ...
	resources: ...
	env: ...
workers:
	foo-worker: ...
cronJobs:
	foo-hourly: ...
```

This allows any combination of workload to be used. E.g., a service could have two deployments (API + UI) or just CronJobs.

In addition, this'd make it easy to add other types of workloads (e.g., daemon set/Job).

**Comment by gleaver** [comment]
@janslow those two manifest snippets look identical - is that intended?

**Comment by clux** [comment]
I've updated @janslow 's comment with what he told me on slack.

**Comment by clux** [comment]
note that the `workloads` property is equivalent to this structure if we vectorised the various ones.

```yaml
name: foo
workloads:
  deployments:
   - name: foo-api
  workers:
  - name: foo-headless-worker
  cronjobs:
  - name: foo-hourly
```

this is going to be easier to pick up in the charts.
interestingly this is almost equivalent to the syntax we already have:

```yaml
workers:
- name: foo-headless-worker
cronjobs:
- name: foo-hourly
```

.. but with the main deployment properties on root.

**Comment by clux** [comment]
If we go down the workloads route, we should consider what properties are tied to the main deployment. A few important ones:

```yaml
replicaCount -> moves into relevant deployment
health/readinessProbe/livenessProbe -> moves into the relevant deployment
autoScaling  -> must move into relevant deployment
configs -> mounted universally
kong -> most be set on the deployment, and must be opitonal
lifecycle -> must move into deployment
resources -> must move into deployment
httpPort -> must move into deployment (if we can't omit it via `ports`)
env -> must move into deployment
```

The only annoying part is how the `httpPort` is tied to the main service's target port and it's duplicated in `ports` vec. We should clean that up.


A move to workloads is certainly doable, but the thing I don't like about this is it's such a huge change that nests almost everything two levels of indentation down (necessitating a bunch of more nuanced merge rules) just for the few services that do need actually need `workers` or `cronjobs`.

In our manifests:
- 4% of services use `cronJobs`
- 5% of services use `workers`

On one hand it would be nice to make the merge rules for the dumb vectorised `workers` and `cronJobs` better, but I wonder if the `workloads` way is the right way to do that.

**Comment by clux** [comment]
So that means the `mainWorkload` or a `primary` key is probably preferable. Although if we have a `primary` key, why do we need to move the nested properties into it? We're still sticking to the shipcat paradigm.

And if we're not moving things down into the `primary` key, why not just have an enumed set of kinds?

**Comment by clux** [comment]
The current most accepted solution is a `workload` key to designate the primary workload (rather than to weirdly use the `kind` key and end up with 2 different kind fields on the k8s object. See linked pr.

**Comment by clux** [comment]
Going with `workload` for now.

**Comment by clux** [comment]
Implemented in 0.144.0
